add optional automatic defmt implementation

agrif opened the pull request:
This PR adds a `defmt` option to the `bitfield` macro, which adds code to implement `defmt::Format`. It also adds info to the README and a few appropriate tests.

Some implementation notes you might want to know:

 * Defmt has [built-in support for bitfields][defmt-bitfield] which leads to a very compact representation, however this method always outputs plain integers for the fields. I chose *not* to use this, and instead format each field separately, in order to take advantage of the `defmt::Format` implementations on any custom types. I feel like this better matches the default `fmt::Debug` implementation. I considered doing both, and making it configurable, but that seemed like too big a change.

  [defmt-bitfield]: https://defmt.ferrous-systems.com/ser-bitfield

 * This macro assumes any type named `u8` is the primitive `u8`, etc. This is also what the official defmt derive macro does, to more efficiently encode these values. It looks like the `bitfield` macro already makes this assumption when figuring out default bit widths, so probably not a big deal.

 * The generated code uses `defmt` to refer to defmt, and not `::defmt`. This is what the defmt crate itself does in its own macros, for [some pretty subtle reasons][defmt-name]. I've added a big comment in the code about this, to hopefully prevent somebody from 'fixing' it.

  [defmt-name]: https://github.com/knurling-rs/defmt/pull/835

Please let me know if you would like any changes, or have any questions!

wrenger commented on src/lib.rs:
```diff
@@ -28,6 +28,7 @@ fn s_err(span: proc_macro2::Span, msg: impl fmt::Display) -> syn::Error {
 /// - `from` to specify a conversion function from repr to the bitfield's integer type
 /// - `into` to specify a conversion function from the bitfield's integer type to repr
 /// - `debug` to disable the `Debug` trait generation
+/// - `defmt` to enable the `defmt::Format` trait generation.
```
There are multiple popular formatting crates for embedded, and usually you only use one of them.
Maybe it would be better to only have the `debug` attribute but with different values for the formatting crates:

```rust
#[bitfield(u16, debug = None)]
struct Foo { /*...*/ }
#[bitfield(u16, debug = Core)] // <- Default
struct Foo { /*...*/ }
#[bitfield(u16, debug = Defmt)]
struct Foo { /*...*/ }
#[bitfield(u16, debug = Ufmt)]
struct Foo { /*...*/ }
// ...
```

agrif replied:
Should `debug = Defmt` still implement `core::fmt::Debug` also? I usually implement both in my crates, simply because `Debug` can be relied upon to almost always exist. If a bitfield ends up buried inside a foreign type I want to look at, I don't really want to have to remove the `Defmt` implementation (and any existing code that uses it) for `Debug` to work.

agrif replied:
Something I just remembered: a lot of crates (mine included) put `defmt` implementations behind a feature flag. Using something like `defmt = cfg!(feature = "defmt")` doesn't work, because the proc macro expects a boolean literal there, and I suspect the same will happen with the enum argument.

Instead, I do something like this:

~~~~
#[cfg_attr(not(feature = "defmt"), bitfield(u16))]
#[cfg_attr(feature = "defmt", bitfield(u16, defmt = true))]
pub struct Register { .. }
~~~~

This is fine, but awkward, and will only get worse if I had more arguments to `bitfield`. Can you think of a better way? Maybe a `debug-feature` argument?

wrenger replied:
Ok if it's a realistic use case to have multiple different formatting implementations then the enum approach doesn't really work. Alternatives are multiple bool attributes or accepting multiple values for the debug attribute. The first is more backwards compatible, so let's just keep it.
For expressions beyond bool literals we would have to evaluate them first.

wrenger commented on src/lib.rs:
```diff
@@ -686,6 +758,7 @@ struct Params {
     from: Option<syn::Path>,
     bits: usize,
     debug: bool,
+    defmt: bool,
```
If we only have the `debug` attribue, an enum (similar to the field `Order`) would be better.

wrenger commented on src/lib.rs:
```diff
@@ -119,6 +121,42 @@ fn bitfield_inner(args: TokenStream, input: TokenStream) -> syn::Result<TokenStr
         quote!()
     };
 
+    let defmt_impl = if defmt {
```
This function is getting quite long (especially if we want to support different formatters).
I would recommend moving the formatting code (including `core::fmt`) into different functions (including the parts inside `Member::debug` / `Member::defmt`).
Something alongside the following would be nice at the top level:

```rust
let debug_impl = implement_debug(debug, name, &members);
```

`debug` is an enum (like `Debug::Defmt`/`Debug::Core`) as explained earlier.

agrif replied:
Do you mean inlining the existing `Member::debug`/`Member::defmt` inside this new `implement_debug` function?

wrenger replied:
Yes, then the format implementation are mostly self contained.

wrenger commented on src/lib.rs:
```diff
@@ -316,6 +356,38 @@ impl Member {
+
+        let ident = &inner.ident;
+        Some((format!("{}: {}", ident, spec), quote!(self.#ident())))
+    }
```
If you use plain `format!`, inlining the variables is a bit more readable. (Only `format_ident!` from quote breaks when you do this.)
```suggestion
        Some((format!("{ident}: {spec}"), quote!(self.#ident())))
```

wrenger replied:
Well, the suggestions are on the wrong line, sorry.

wrenger commented on tests/test.rs:
```diff
@@ -133,6 +133,107 @@ fn debug() {
 }
 
+// a dummy defmt logger and timestamp implementation, for testing
+#[cfg(test)]
```
Isn't this redundant? The whole file is only compiled for `test`.

agrif replied:
Yes -- this is cruft from an earlier revision where I was trying to get a single dummy implementation that would work for all tests and doctests, but doctests especially resist conditional compilation like this.

wrenger commented on src/lib.rs:
```diff
@@ -316,6 +356,38 @@ impl Member {
+        // get the type name so we can use more efficient defmt formats
+        // if it's a primitive
+        if let syn::Type::Path(syn::TypePath { ref path, .. }) = inner.ty {
+            if let Some(ident) = path.get_ident() {
```
I usually don't use the `ref` keyword anymore and just put an `&` before `inner.ty`.

```suggestion
        if let syn::Type::Path(syn::TypePath { path, .. }) = &inner.ty {
```

wrenger reviewed (commented):
Your changes are quite good. Still, I would change a few parts to make it easier to add new formatting/logging crates in the future, if the need arises.

agrif commented:
I'll start on these changes, I just have a few clarifying questions first...

agrif commented:
Ok, I removed the extra `#[cfg(test)]` and moved the debug implementations into `implement_debug`. I've left these as separate commits for the sake of discussion, but of course you can squash them (or not) as you please.

If you're keeping the separate boolean flags, some other projects use e.g. `#[bitfield(u16, defmt = true)]` to unconditionally implement defmt, and `#[bitfield(u16, defmt = "flag")]` to conditionally implement it behind a feature flag. I'd be happy to add that (for debug also??), but since it mixes argument types I know not everybody is happy about that.

wrenger commented:
Ok, thanks for the changes. I merge this.

wrenger commented:
> If you're keeping the separate boolean flags, some other projects use e.g. `#[bitfield(u16, defmt = true)]` to unconditionally implement defmt, and `#[bitfield(u16, defmt = "flag")]` to conditionally implement it behind a feature flag. I'd be happy to add that (for debug also??), but since it mixes argument types I know not everybody is happy about that.

Let's do this in another PR, as accessing features inside proc macros is a bit tedious.

agrif commented:
> Let's do this in another PR, as accessing features inside proc macros is a bit tedious.

Oh? It's more complicated than adding `#[cfg(feature = #feature_name)]` before the impl inside the quote?

wrenger commented:
Ive extended this to support arbritrary `cfg` attributes in afa7fd8774ad70ec68b6c731794afc3bac729d62
